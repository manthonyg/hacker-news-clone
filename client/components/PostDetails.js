/* eslint-disable no-console */
import React, { useState, useEffect } from 'react';
import queryString from 'query-string';
import moment from 'moment';
import { Link } from 'react-router-dom';
import styled from 'styled-components';
import { fetchComments, fetchItem } from '../utils/api';
import Comment from './Comment';
import Flex from './common/Flex';
import UserInfoSkeleton from './UserInfoSkeleton';
import Heading from './common/Heading';
import PostSkeleton from './PostSkeleton';

const StyledLink = styled(Link)`
  text-decoration: underline;
  color: #bb86fc;
  font-weight: 800;
`;

function SinglePost() {
  // eslint-disable-next-line no-restricted-globals
  const { id } = queryString.parse(location.search);

  const [comments, setComments] = useState([]);
  const [post, setPost] = useState(null);
  const [isLoading, setIsLoading] = useState(false);

  useEffect(() => {
    setIsLoading(true);
    fetchItem(id)
      .then(userpost => {
        setPost(userpost);
        return fetchComments(userpost.kids || []);
      })
      .then(usercomments => {
        setComments(usercomments);
        setIsLoading(false);
      })
      .catch(error => {
        console.warn('There was a problem gathering these resources', error);
      });
  }, []);

  return (
    <>
      {isLoading ? (
        <>
          <PostSkeleton numberOfSkeletons={1} noEmoji />
          <UserInfoSkeleton numberOfSkeletons={3} />
        </>
      ) : (
        <>
          {post && (
            <>
              <Heading h4 noMargin>
                {post.title}
              </Heading>
              <Heading h5>
                by <StyledLink to={`/user?id=${post.by}`}>{post.by}</StyledLink>{' '}
                at {moment.unix(post.time).format('YYYY-MM-DD')}
                title: {post.title}
              </Heading>
            </>
          )}
          <Flex>
            {comments && comments.length ? (
              comments.map(comment => (
                <Comment
                  key={comment.id}
                  id={comment.id}
                  text={comment.text}
                  by={comment.by}
                  time={comment.time}
                />
              ))
            ) : (
              <Heading>No Comments</Heading>
            )}
          </Flex>
        </>
      )}
    </>
  );
}

export default SinglePost;
