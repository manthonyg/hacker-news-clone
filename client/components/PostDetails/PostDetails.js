/* eslint-disable no-console */
import React, { useState, useEffect, useRef } from "react";
import queryString from "query-string";
import Loader from "../common/Loader/Loader";
import { fetchComments, fetchItem } from "../../utils/api";
import Flex from "../common/Flex/Flex";
import Heading from "../common/Heading/Heading";
import PostSkeleton from "../Posts/PostSkeleton";
import Post from "../Post/Post";
import CommentList from "../CommentList/CommentList";
import InfiniteScroll from "react-infinite-scroll-component";
import ScrollArrow from "../ScrollArrow/ScrollArrow";

function PostDetails() {
  const INFINITE_SCROLL_FETCH_AMOUNT = 10;
  const { id } = queryString.parse(location.search);
  const [comments, setComments] = useState([]);
  const [commentIds, setCommentIds] = useState([]);
  const [post, setPost] = useState(null);
  const [isLoading, setIsLoading] = useState(false);

  const componentIsMounted = useRef(true);

  useEffect(() => {
    if (componentIsMounted.current) {
      setIsLoading(true);
      fetchItem(id)
        .then((post) => {
          setPost(post);
          setCommentIds(post?.kids);
          return fetchComments(
            post?.kids.slice(0, INFINITE_SCROLL_FETCH_AMOUNT) || []
          );
        })
        .then((comments) => {
          setComments(comments);
          setIsLoading(false);
        })
        .catch((error) => {
          console.warn("There was a problem gathering these resources", error);
        });
    }
    return () => {
      componentIsMounted.current = false;
    };
  }, []);

  const handleInfiniteScroll = () => {
    const currentCommentLength = comments.length;
    fetchComments(
      commentIds.slice(
        currentCommentLength - 1,
        currentCommentLength + INFINITE_SCROLL_FETCH_AMOUNT
      )
    ).then((response) => setComments(comments.concat(response)));
  };

  return (
    <>
      {isLoading ? (
        <Flex>
          <Heading h5>viewing:</Heading>
          <PostSkeleton numberOfSkeletons={1} />
          <Heading h5>comments: (...)</Heading>
          <PostSkeleton numberOfSkeletons={10} noEmoji />
        </Flex>
      ) : (
        <>
          <Heading h5>viewing:</Heading>
          {post && <Post post={post} />}
          {comments && (
            <Heading h5 isSticky>
              comments ({comments?.length} of {commentIds?.length}):
            </Heading>
          )}

          <InfiniteScroll
            style={{ overflow: "auto" }}
            dataLength={comments?.length}
            next={handleInfiniteScroll}
            hasMore={comments?.length <= commentIds?.length - 1}
            scrollThreshold={0.95}
            loader={
              <>
                <Heading h5>Loading comments...</Heading>
                <Loader />
              </>
            }
            endMessage={<Heading h4>no more comments</Heading>}
          >
            <CommentList comments={comments} />
          </InfiniteScroll>
          <ScrollArrow />
        </>
      )}
    </>
  );
}

export default PostDetails;
