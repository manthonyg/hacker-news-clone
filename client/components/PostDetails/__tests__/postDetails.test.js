import React from "react";
import renderer from "react-test-renderer";
import PostDetails from "../PostDetails";

it("renders correctly with all props", () => {
  const requiredProps = {};
  const tree = renderer
    .create(
      <>
        <PostDetails {...requiredProps} />
      </>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
