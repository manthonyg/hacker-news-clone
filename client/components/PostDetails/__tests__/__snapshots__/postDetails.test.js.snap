// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with all props 1`] = `
Array [
  <h5
    className="sc-fzozJi ekWdoI"
  >
    viewing:
  </h5>,
  <div
    className="sc-AxiKw dGHlTc"
  >
    <h5
      className="sc-fzozJi ekWdoI"
    >
      comments (
      0
       of 
      0
      ):
    </h5>
    <div
      className="infinite-scroll-component__outerdiv"
      style={Object {}}
    >
      <div
        className="infinite-scroll-component "
        style={
          Object {
            "WebkitOverflowScrolling": "touch",
            "height": "auto",
            "overflow": "auto",
          }
        }
      >
        <h1
          className="sc-AxgMl igiwKw"
        >
          no more comments
        </h1>
      </div>
    </div>
  </div>,
]
`;
